### ISSUE TITLE
Parallelize script to read in bundles and populate an S3 bucket with de-duped files.

### ISSUE THREAD

**Issue by maniarathi** [HumanCellAtlas/dcp-prototype#84]
As part of the effort to migrate DCP 1.0 data into DCP 2.0 and convert the metadata schema into the DCP 2.0 metadata schema, the end-to-end conversion occurs in two steps:

1) First, the full set of bundles for a single project is downloaded to an EC2 instance following which a script reads, each file in the bundle, determines whether it has been previously copied to an S3 bucket via checksum, and if not, copies the file over to an S3 bucket in the DCP 2.0 AWS account. The files are "flattened" which means that it no longer only exists in its bundle directory so that duplicates may be removed. This process also allows us to retain a lossless copy of all DCP 1.0 data.

2) Once all the data for a project has been copied and de-duped into a single folder in an S3 bucket, a script reads each file, and if it is related to constructing the metadata graph for the project (these are all the .json files), reads the entity into a Python object. Once all entities have been read, all `links.json` files are read to construct the edges between the Python objects that have been created. Following this, the metadata graph is transformed into entities that can be directly inputted into the Ledger using the DCP 2.0 metadata schema.

This ticket refers specifically to part 1 of the above process. This script has no parallelization and is disconnected from the "download_bundle" portion so the goal of this ticket is to speed-up the script as well as connect it to the downloading of the bundles.

**Comment by maniarathi** [comment]
Hey @MDunitz, just wondering what the status of this ticket is; I know that you managed to copy over all the data to an S3 bucket but what's the status on the script getting checked in? Doesn't need to be perfect since we hopefully don't need to do it again, but probably safe to have it around just in case.

Also, just curious, what were the changes you made the script to make it run faster? Just threaded? And how quickly does it run?

Question for @ambrosejcarr: the S3 uri for most of the FASTQ/BAM/etc. files use the FQID as the filename instead of the old SRR1234567_0.fastq.gz. FQID I realize is a much longer filename and I realized that the link will be exposed in the browser directly to download. Do you have a preference whether we just use the old name ("SRR1234567_0.fastq.gz") or the fqid ("0e7b2404-c529-43fb-a55d-199a76caa860.2019-08-01T031948.896Z.fastq.gz")?

**Comment by parthshahva** [comment]
In my opinion, we need to use the old name. The fqid is a left over from DCP 1.0 DSS data model that has no bearing on 2.0. It will only serve to confuse users even if its packaged up

**Comment by ambrosejcarr** [comment]
@liaprins-czi has started some work on file naming in #75 that @jahilton is consulting on. The decision there should govern file naming. If that's blocking, using a placeholder in the interim is acceptable. 

While the published artifact will contain files and should have informative names, the browser will not be directly serving those files initially, it will be bundling fastq and bam files by metadata categories and will be slicing up matrix files into smaller pieces to match the same metadata categorization.

**Comment by maniarathi** [comment]
@MDunitz Given the above discussion, would you be able to re-transfer/copy the files please using the original names instead of the fqid. I suppose in retrospect, it also makes it easier to run the second half of the script that Brian was working on, since it implies the names of files based on the `filename` attribute of `sequence_file.json` or `analysis_file.json`. Likewise for the metadata files that use the filename to determine what kind of entity it is. Apologies for the re-run!

Thank you!!

**Comment by MDunitz** [comment]
So currently the s3uri for the data files is a concatenation of the hashes for the file (to simplify copying over the data and avoid pulling large data files more than once). 
The file name for metadata (json) files is the fqid because I thought that would make it simpler to connect to links.json (which refers to files by their uuids).  
It’s fairly simple to rename the files (now or when we read them in as python objects) if we prefer a different naming structure. 
Due to the size of the data files I think it’s probably worth it to continue to name them based on their hash to avoid storing multiple copies under different names. 
There may be a way to do sym links in s3 or change the name of the file when it’s downloaded to make it more clear for users. 

@maniarathi re status I’m mainly waiting on a rereview. 
Copying the data files takes ~24hrs for ~20tb. The metadata files take about 12hrs. They are currently available in the s3 bucket but I think it might be faster to read them into the db from the ec2 instance

**Comment by MDunitz** [comment]
Sorry saw your comment after I made mine. Happy to update the script and rerun

**Comment by maniarathi** [comment]
Thanks @MDunitz! Oops, sorry, will re-review today :) 

Are there duplicates of the data files? I thought the only ones that were duplicates were some of the zarr cell or gene attribute name files which are small and easy to de-dup via checksum. The large files, like the BAMs and the FASTQs are all unique?

**Comment by MDunitz** [comment]
I could be wrong but I thought there was concern about the same data file being referenced with a different uuid/name in different bundles. 
I used the bundle manifests to ensure I only pulled data files associated with the 29 projects that are currently available in prod. I also wasn’t confident the file names in the bundle manifest were unique across all projects.

**Comment by maniarathi** [comment]
Hey @MDunitz, could you merge the PR associated with this ticket and close please? Thanks!

**Comment by MDunitz** [comment]
@maniarathi can you accept the changes you requested and lgtm the pr first?

**Comment by maniarathi** [comment]
Oops! Done!
